vllm/v1/worker/neuron_model_runner.py

```diff
@@ -0,0 +1,993 @@
+                if start_pos + num_encoder_tokens <= num_computed_tokens:
+                    # The encoder output is already processed and stored
+                    # in the decoder's KV cache.
+                    continue
+
+                start_idx = max(num_computed_tokens - start_pos, 0)
+                end_idx = min(
+                    num_computed_tokens - start_pos + num_scheduled_tokens,
+                    num_encoder_tokens)
+                assert start_idx < end_idx
+                assert req_id in self.encoder_cache
+                assert i in self.encoder_cache[req_id]
+                encoder_output = self.encoder_cache[req_id][i]
+                encoder_outputs.append(encoder_output[start_idx:end_idx])
+        return encoder_outputs
+
+    @torch.inference_mode()
+    def execute_model(
+        self,
+        scheduler_output: "SchedulerOutput",
+        intermediate_tensors: Optional[IntermediateTensors] = None,
+    ) -> ModelRunnerOutput:
+        self._update_states(scheduler_output)
+
+        if self.is_multimodal_model:
+            # Run the multimodal encoder if any.
+            self._execute_encoder(scheduler_output)
+            encoder_outputs = self._gather_encoder_outputs(scheduler_output)
+        else:
+            encoder_outputs = []
+
+        # Prepare the decoder inputs.
+        attn_metadata, logits_indices = self._prepare_inputs(scheduler_output)
+        num_scheduled_tokens = scheduler_output.total_num_scheduled_tokens
+        num_input_tokens = self._get_padded_batch_size(num_scheduled_tokens)
+
+        attn_metadata.num_input_tokens = num_input_tokens
+
+        if self.is_multimodal_model:
+            # NOTE(woosuk): To unify token ids and soft tokens (vision
+            # embeddings), we always use embeddings (rather than token ids)
+            # as input to the multimodal model, even when the input is text.
+            input_ids = self.input_ids[:num_scheduled_tokens]
+            if encoder_outputs:
+                inputs_embeds = self.model.get_input_embeddings(
+                    input_ids, encoder_outputs)
+            else:
+                inputs_embeds = self.model.get_input_embeddings(input_ids)
+            # TODO(woosuk): Avoid the copy. Optimize.
+            self.inputs_embeds[:num_scheduled_tokens].copy_(inputs_embeds)
+            inputs_embeds = self.inputs_embeds[:num_input_tokens]
+            input_ids = None
+        else:
+            # For text-only models, we use token ids as input.
+            input_ids = self.input_ids[:num_input_tokens]
+            inputs_embeds = None
+        if self.uses_mrope:
+            positions = self.mrope_positions[:, :num_input_tokens]
+        else:
+            positions = self.positions[:num_input_tokens]
+
+        # Run the decoder.
+        # Use persistent buffers for CUDA graphs.
+        with set_forward_context(attn_metadata, self.vllm_config):
+            hidden_states = self.model(
```

Comment:
In the original PoC with this model runner class, I had to make a couple changes to the underlying model arch class (in my case Llama). Definitely don't want any neuron-specific logic in the model classes, but I had not found a way yet to avoid it 

Specifically, these changes were:
1. Remove the call to `qkv.split` in `LlamaAttention`, as this resulted in issues during neuron compilation.
2. Move the `lm_head` to CPU to keep logits processing and sampling on CPU. 

I was curious if/how you found a way to eliminate the need for these changes in this PR?

Reply:
>Remove the call to qkv.split in LlamaAttention, as this resulted in issues during neuron compilation.

We can workaround the qkv.split issue with the following change in `platforms/neuron.py`
```
        # Disable functionalization on xla, due to https://github.com/pytorch/xla/issues/8640
        os.environ["XLA_DISABLE_FUNCTIONALIZATION"] = "0"
```

> Move the lm_head to CPU to keep logits processing and sampling on CPU.

We have moved this hack to `llama.py` into `neuron_model_runner.py` with
```
        if hasattr(self.model, "lm_head"):
            # Compute logits on CPU
            self.model.lm_head = self.model.lm_head.cpu()
 ```